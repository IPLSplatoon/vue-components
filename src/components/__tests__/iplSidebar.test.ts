import IplSidebar from '../iplSidebar.vue';
import { flushPromises, mount } from '@vue/test-utils';

describe('IplSidebar', () => {
    // https://github.com/jsdom/jsdom/issues/3294
    beforeAll(() => {
        HTMLDialogElement.prototype.showModal = jest.fn();
        HTMLDialogElement.prototype.close = jest.fn();
        Element.prototype.getAnimations = jest.fn().mockReturnValue([]);
    });

    it('matches snapshot', () => {
        expect(mount(IplSidebar, {
            props: { isOpen: false },
            slots: {
                default: () => 'test sidebar content'
            }
        }).html()).toMatchSnapshot();
    });

    it('opens the modal if mounted as open', () => {
        const wrapper = mount(IplSidebar, { props: { isOpen: true } });
        expect(wrapper.get('dialog').element.showModal).toHaveBeenCalled();
    });

    it('emits an event when the modal is closed', () => {
        const wrapper = mount(IplSidebar, { props: { isOpen: true } });

        wrapper.get('dialog').element.dispatchEvent(new Event('close'));

        expect(wrapper.emitted('update:isOpen')).toEqual([[false]]);
    });

    it('closes the dialog when directly clicked', async () => {
        const wrapper = mount(IplSidebar, { props: { isOpen: true } });

        const dialog = wrapper.get('dialog');
        const event = new Event('click');
        Object.defineProperty(event, 'target', { value: dialog.element });
        dialog.element.dispatchEvent(event);
        await flushPromises();

        expect(dialog.element.close).toHaveBeenCalled();
        expect(dialog.element.hasAttribute('inert')).toEqual(true);
    });

    it('does not close the dialog when directly clicked and persistent', async () => {
        const wrapper = mount(IplSidebar, { props: { isOpen: true, persistent: true } });

        const dialog = wrapper.get('dialog');
        const event = new Event('click');
        Object.defineProperty(event, 'target', { value: dialog.element });
        dialog.element.dispatchEvent(event);
        await flushPromises();

        expect(dialog.element.close).not.toHaveBeenCalled();
        expect(dialog.element.hasAttribute('inert')).toEqual(false);
    });

    it('does not close the dialog when clicking on a child of the dialog', async () => {
        const wrapper = mount(IplSidebar, { props: { isOpen: true } });

        const dialog = wrapper.get('dialog');
        const event = new Event('click');
        Object.defineProperty(event, 'target', { value: document.createElement('div') });
        dialog.element.dispatchEvent(event);
        await flushPromises();

        expect(dialog.element.close).not.toHaveBeenCalled();
    });

    it('closes the dialog when cancelled', async () => {
        const wrapper = mount(IplSidebar, { props: { isOpen: true } });

        const dialog = wrapper.get('dialog');
        const event = new Event('cancel');
        jest.spyOn(event, 'preventDefault');
        dialog.element.dispatchEvent(event);
        await flushPromises();

        expect(dialog.element.close).toHaveBeenCalled();
        expect(dialog.element.hasAttribute('inert')).toEqual(true);
        expect(event.preventDefault).toHaveBeenCalled();
    });
});
