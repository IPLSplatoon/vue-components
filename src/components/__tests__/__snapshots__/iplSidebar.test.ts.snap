// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IplSidebar matches snapshot when hidden 1`] = `
"<transition-stub name="sidebar-transition" duration="250" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;

exports[`IplSidebar matches snapshot when visible 1`] = `
"<transition-stub name="sidebar-transition" duration="250" appear="false" persisted="false" css="true">
  <div class="ipl-sidebar__wrapper">
    <div class="background"></div>
    <div class="content"></div>
  </div>
</transition-stub>"
`;
